core/src/main/java/org/infinispan/remoting/transport/MultiTargetRequest.java

```diff
@@ -0,0 +1,164 @@
+            if (missingResponses <= 0) {
+               // The request is completed, nothing to do
+               return;
+            }
+
+            boolean validSender = false;
+            for (int i = 0; i < targets.length; i++) {
+               Address target = targets[i];
+               if (target != null && target.equals(sender)) {
+                  validSender = true;
+                  targets[i] = null;
+                  missingResponses--;
+                  break;
+               }
+            }
+
+            if (!validSender) {
+               throw new IllegalStateException(
+                     "Received invalid response for request " + requestId + " from " + sender);
+            }
+
+            result = responseCollector.addResponse(sender, response);
+            if (result != null) {
+               isDone = true;
+            } else if (missingResponses <= 0) {
+               isDone = true;
+               result = responseCollector.finish();
+            }
+         }
+
+         // Complete the request outside the lock, in case it has to run blocking callbacks
+         if (isDone) {
+            complete(result);
+         }
+      } catch (Throwable t) {
+         completeExceptionally(t);
+      }
+   }
+
+   @Override
+   public void onNewView(Set<Address> members) {
+      try {
+         boolean isDone = false;
+         T result = null;
+         synchronized (responseCollector) {
+            if (missingResponses <= 0) {
+               // The request is completed, must not modify ResponseObject.
+               return;
+            }
+            for (int i = 0; i < targets.length; i++) {
+               Address target = targets[i];
+               if (target != null && !members.contains(target)) {
+                  targets[i] = null;
+                  missingResponses--;
+                  result = responseCollector.addResponse(target, CacheNotFoundResponse.INSTANCE);
+                  if (result != null) {
+                     isDone = true;
+                     break;
+                  }
+               }
+            }
+
+            // No more targets remaining
+            if (!isDone && missingResponses <= 0) {
+               result = responseCollector.finish();
+               isDone = true;
+            }
+         }
+
+         // Complete the request outside the lock, in case it has to run blocking callbacks
+         if (isDone) {
+            complete(result);
+         }
+      } catch (Throwable t) {
+         completeExceptionally(t);
+      }
+   }
+
+   @Override
+   public void onTimeout() {
+      synchronized (responseCollector) {
+         if (missingResponses <= 0) {
+            // The request is already completed.
+            return;
+         }
+         // Don't add more responses to the collector after this
+         this.missingResponses = 0;
+      }
+
+      Collection<Address> targetsWithoutResponses = Arrays.stream(targets)
```

Comment:
Done... this is a leftover from the time the `ResponseCollector` was also dealing with timeouts.